### DISCUSSION TITLE
Compiling contract code in tests

### DISCUSSION THREAD

**Discussion by 0xGh** [ConsenSys-archive/truffle#6164] [General Q&A] [answered]
Hi there, I want to compile a smart contract code (which has a couple of dependencies) in a test. 

The reason for this is that the contract code is a template. So I would like to do something like this:

```js
it('works', async () => {
   const CompiledContract = await compile(getSource(variables));
   const instance = await CompiledContract.new();
   // ..
})
```

Any pointers?

**Comment by 0xGh** [comment, marked as answer]
Came up with this solution:

```js
const path = require('path');
const fs = require('fs');

const TruffleCompile = require('@truffle/workflow-compile').default;
const TruffleConfig = require('@truffle/config');
const TruffleContract = require('@truffle/contract');

const MyContractSource = require('fs').readFileSync('./contracts/MyContract.sol', 'utf-8');

let truffleConfig = null;
async function compile(source, fileName = 'TestContract') {
  const config = truffleConfig || TruffleConfig.detect();
  if (!truffleConfig) {
    truffleConfig = config;
  }

  const outFile = path.join('contracts', 'mocks', fileName + '.sol');
  fs.writeFileSync(outFile, source);

  if (config._ && config._.length > 0) {
    config.paths = config._.map(specifiedPath => {
      const absolutePath = path.resolve(config.working_directory, specifiedPath);

      if (fs.existsSync(absolutePath)) {
        return absolutePath;
      } else {
        return specifiedPath;
      }
    });
  }

  const compilationOutput = await TruffleCompile.compile(config);

  fs.unlinkSync(outFile);

  const contractArtifact = compilationOutput.compilations[0].contracts.find(
    ({ contractName }) => contractName == fileName,
  );

  if (contractArtifact == null) {
    throw new Error('Contract compiled but not found');
  }

  const contract = TruffleContract(
    contractArtifact.bytecode == 'string'
      ? contractArtifact
      : require('@truffle/compile-common').Shims.NewToLegacy.forContract(contractArtifact),
  );

  contract.setProvider(web3.currentProvider);

  return contract;
}
```

Probably this solution is not the best because it saves the contract on disk but I am sure it can be improved further later on.

**Reply by gnidan** [reply]
Whoa you basically got the exact answer I was preparing to write up :) 

I'm not sure you need to put the contract on disk... you'll need the `NewToLegacy` shim definitely, but you should just be able to get the `WorkflowCompileResult` and massage it for use by the @truffle/contract constructor

**Reply by 0xGh** [reply]
Hi @gnidan thanks for chiming in! Isn't `WorkflowCompile.compile` scanning the contracts folder to find the contracts that needs compilation? Or do you have a `compile` version that takes the source contract as input (string), gets the dependencies from disk and compiles my contract with necessary dependencies?

**Reply by gnidan** [reply]
> Hi @gnidan thanks for chiming in! Isn't WorkflowCompile scanning the contracts folder to find the contracts that needs compilation? Or do you have a compile version that takes the source contract as input (string), gets the dependencies from disk and compiles my contract with necessary dependencies?

Oh yeah, I think you want `Compile.sources()`. It's poorly typed, but hopefully this helps: https://github.com/trufflesuite/truffle/blob/develop/packages/compile-common/src/types.ts#L79 (you might want to see the corresponding implementation in @truffle/workflow-compile or @truffle/compile-solidity, or check out @truffle/fetch-and-compile to see how it does programmatic compilation of sources.)
